### ISSUE TITLE
iptables rules causing network unreachable

### ISSUE THREAD

**Issue by Prodian0013** [kubernetes/kubernetes#83063]
**What happened*:
Pods on 10.244.0.0/16 network can reach each other but cannot reach the 10.96 network or node network.

`E0923 21:25:30.881753       1 reflector.go:134] github.com/coredns/coredns/plugin/kubernetes/controller.go:317: Failed to list *v1.Endpoints: Get https://10.96.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0: dial tcp 10.96.0.1:443: connect: network is unreachable`

`NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP   8m35s`

`dnstools# nc -v 10.96.0.1:443
nc: 10.96.0.1:443 (10.96.0.1:443): Network unreachable
`

> [root@server1 ~]# iptables -L
> Chain INPUT (policy ACCEPT)
> target     prot opt source               destination
> KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
> KUBE-EXTERNAL-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes externally-visible service portals */
> KUBE-FIREWALL  all  --  anywhere             anywhere
> 
> Chain FORWARD (policy ACCEPT)
> target     prot opt source               destination
> KUBE-FORWARD  all  --  anywhere             anywhere             /* kubernetes forwarding rules */
> KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
> ACCEPT     all  --  server1.lan/16       anywhere
> ACCEPT     all  --  anywhere             server1.lan/16
> 
> Chain OUTPUT (policy ACCEPT)
> target     prot opt source               destination
> KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
> KUBE-FIREWALL  all  --  anywhere             anywhere
> 
> Chain KUBE-EXTERNAL-SERVICES (1 references)
> target     prot opt source               destination
> 
> Chain KUBE-FIREWALL (2 references)
> target     prot opt source               destination
> DROP       all  --  anywhere             anywhere             /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000
> 
> Chain KUBE-FORWARD (1 references)
> target     prot opt source               destination
> DROP       all  --  anywhere             anywhere             ctstate INVALID
> ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding rules */ mark match 0x4000/0x4000
> ACCEPT     all  --  server1.lan/16       anywhere             /* kubernetes forwarding conntrack pod source rule */ ctstate RELATED,ESTABLISHED
> ACCEPT     all  --  anywhere             server1.lan/16       /* kubernetes forwarding conntrack pod destination rule */ ctstate RELATED,ESTABLISHED
> 
> Chain KUBE-SERVICES (3 references)
> target     prot opt source               destination
> REJECT     tcp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:dns-tcp has no endpoints */ tcp dpt:domain reject-with icmp-port-unreachable
> REJECT     tcp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:metrics has no endpoints */ tcp dpt:9153 reject-with icmp-port-unreachable
> REJECT     udp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:dns has no endpoints */ udp dpt:domain reject-with icmp-port-unreachable

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
kubadm init --pod-network-cidr=10.244.0.0/16

**Anything else we need to know?**:
Running CRI-O.
selinux is enforcing and firewalld is running.
Kubernetes 1.13 does not have this issue (our current production version).
It appears somethings have changed regarding selinux and iptables from 1.13 to 1.15.

**Environment**:
- Kubernetes version (use `kubectl version`): Kubernetes v1.15.3
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): Red Hat Enterprise Linux Server release 7.7 (Maipo)
- Kernel (e.g. `uname -a`): inux server1.lan 3.10.0-1062.1.1.el7.x86_64 #1 SMP Tue Aug 13 18:39:59 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: kubeadm
- Network plugin and version (if this is a network-related bug): containernetworking-plugins-0.8.1 and Flannel v0.11.0
- Others: crio version 1.15.1-2.el7

**Comment by Prodian0013** [comment]
After rebooting the node iptables changes and pod networking works.

> [root@server1 ~]# iptables -L
> Chain INPUT (policy ACCEPT)
> target     prot opt source               destination
> KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
> KUBE-EXTERNAL-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes externally-visible service portals */
> KUBE-FIREWALL  all  --  anywhere             anywhere
> 
> Chain FORWARD (policy ACCEPT)
> target     prot opt source               destination
> KUBE-FORWARD  all  --  anywhere             anywhere             /* kubernetes forwarding rules */
> KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
> ACCEPT     all  --  server1.lan/16       anywhere
> ACCEPT     all  --  anywhere             server1.lan/16
> 
> Chain OUTPUT (policy ACCEPT)
> target     prot opt source               destination
> KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
> KUBE-FIREWALL  all  --  anywhere             anywhere
> 
> Chain KUBE-EXTERNAL-SERVICES (1 references)
> target     prot opt source               destination
> 
> Chain KUBE-FIREWALL (2 references)
> target     prot opt source               destination
> DROP       all  --  anywhere             anywhere             /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000
> 
> Chain KUBE-FORWARD (1 references)
> target     prot opt source               destination
> DROP       all  --  anywhere             anywhere             ctstate INVALID
> ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding rules */ mark match 0x4000/0x4000
> ACCEPT     all  --  server1.lan/16       anywhere             /* kubernetes forwarding conntrack pod source rule */ ctstate RELATED,ESTABLISHED
> ACCEPT     all  --  anywhere             server1.lan/16       /* kubernetes forwarding conntrack pod destination rule */ ctstate RELATED,ESTABLISHED
> 
> Chain KUBE-SERVICES (3 references)
> target     prot opt source               destination

**Comment by athenabot** [comment]
/triage unresolved

Comment `/remove-triage unresolved` when the issue is assessed and confirmed.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by aojea** [comment]
hi @Prodian0013 , any clue that what could be the problem?
should we keep it open?

**Comment by Prodian0013** [comment]
> hi @Prodian0013 , any clue that what could be the problem?
> should we keep it open?

yes I was able to reproduce the issue but still dont know whats causing it.

**Comment by Prodian0013** [comment]
the rules in question are:

> Chain KUBE-SERVICES (3 references)
> target     prot opt source               destination
> REJECT     tcp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:metrics has no endpoints */ tcp dpt:9153 reject-with icmp-port-unreachable
> REJECT     udp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:dns has no endpoints */ udp dpt:domain reject-with icmp-port-unreachable
> REJECT     tcp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:dns-tcp has no endpoints */ tcp dpt:domain reject-with icmp-port-unreachable

After a fresh kubeadm init these rules are added to iptables.  These rules appear to be added by kube-proxy.  I tried to following to remove the rules but the rules are repopulated when the pods come back up.

`systemctl stop kubelet;
crictl stop $(crictl ps -q);
crictl rm $(crictl ps -qa);
iptables -F;
systemctl restart firewalld;
systemctl start kubelet;`

After rebooting the node the rules still exist after the pods come back up the rules eventually are removed then pod networking starts working and coredns comes up successfully.

**Comment by Prodian0013** [comment]
Im seeing this issue with versions `1.14.7` and `1.15.3`.

No issue with `1.13.11`

**Comment by aojea** [comment]
@Prodian0013 
> kube-system/kube-dns:dns has no endpoints

it seems the kube-dns service has no endpoints, and kube-proxy install rules to deny the traffic directed to that service.
Can you try to find out what happen with the kube-dns endpoints services?

**Comment by Prodian0013** [comment]
It appers on after kubeadm init the kube-dns endpoint is showing `Addresses:          <none>`
> [root@controller ~]# kubectl describe ep kube-dns -n kube-system
> Name:         kube-dns
> Namespace:    kube-system
> Labels:       k8s-app=kube-dns
>               kubernetes.io/cluster-service=true
>               kubernetes.io/name=KubeDNS
> Annotations:  endpoints.kubernetes.io/last-change-trigger-time: 2019-09-27T13:48:06Z
> Subsets:
>   Addresses:          <none>
>   NotReadyAddresses:  10.244.0.10,10.244.0.9
>   Ports:
>     Name     Port  Protocol
>     ----     ----  --------
>     dns      53    UDP
>     dns-tcp  53    TCP
>     metrics  9153  TCP
> 
> Events:  <none>

**Comment by aojea** [comment]
ok, seems we've found the problem, I think that in 1.14 you should be using CoreDNS instead of KubeDND because [CoreDNS went GA in 1.11](https://kubernetes.io/blog/2018/07/10/coredns-ga-for-kubernetes-cluster-dns/) and it's probably that KubeDNS was deprecated after 3 cycles.

The kube-proxy behavior is correct, please check that you are doing the upgrade correctly https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/ and using CoreDNS instead of KubeDNS for 1.14+

/cc: @neolit123 
/remove triage/unresolved

**Comment by Prodian0013** [comment]
Heres the kubeadm init logs: https://gist.github.com/Prodian0013/e7440857431a99c025be9c35d903f0e3

By default 1.15.3 is using coreDNS

**Comment by Prodian0013** [comment]
It appears the coreDNS service is still called kube-dns.  This should probably be changed to avoid further confusion.

https://github.com/kubernetes/kubernetes/blob/898042007cf3063794398fd2954e1b5b0ee813aa/cluster/addons/dns/coredns/coredns.yaml.base#L176

Im still unsure why the service is not coming up.

**Comment by aojea** [comment]
Yeah, seem my theory was wrong :sweat_smile: 

> NotReadyAddresses: 10.244.0.10,10.244.0.9

Try to troubleshoot the service and the pods, this is what I see 

```
kubectl get pods -n kube-system -l k8s-app=kube-dns
NAME                       READY   STATUS    RESTARTS   AGE
coredns-5c98db65d4-6ctb5   1/1     Running   0          27h
coredns-5c98db65d4-mbxsr   1/1     Running   0          27h
```

```
kubectl describe services kube-dns -n kube-systemName:              kube-dns
Namespace:         kube-system
Labels:            k8s-app=kube-dns
                   kubernetes.io/cluster-service=true
                   kubernetes.io/name=KubeDNS
Annotations:       prometheus.io/port: 9153
                   prometheus.io/scrape: true
Selector:          k8s-app=kube-dns
Type:              ClusterIP
IP:                10.96.0.10
Port:              dns  53/UDP
TargetPort:        53/UDP
Endpoints:         10.244.0.2:53,10.244.0.3:53
Port:              dns-tcp  53/TCP
TargetPort:        53/TCP
Endpoints:         10.244.0.2:53,10.244.0.3:53
Port:              metrics  9153/TCP
TargetPort:        9153/TCP
Endpoints:         10.244.0.2:9153,10.244.0.3:9153
Session Affinity:  None
Events:            <none>
```

**Comment by aojea** [comment]
@Prodian0013 were you able to find out what's causing the issues with the services?

**Comment by aojea** [comment]
@Prodian0013 can you details the steps that you are following, the versions of the OS and kubernetes components and the number of nodees to try to reproduce it?

**Comment by Prodian0013** [comment]
OS: RHEL 7.7
Kubernetes: 1.15.2
CRIO: 1.15.2
Flannel CNI
Single master node

Heres a couple scripts I threw together that will reproduce the issue just for testing purposes:
Setup Node: https://gist.github.com/Prodian0013/87553254f7376a338de60c20531be252
Install K8S: https://gist.github.com/Prodian0013/e5c4a6883bce042dc8bb06f1217c53a7

**Comment by aojea** [comment]
if you are using only one node with kubeadm you have to taint it so it can schedule pods on it.

> Control plane node isolation
> By default, your cluster will not schedule pods on the control-plane node for security reasons. If you want to be able to schedule pods on the control-plane node, e.g. for a single-machine Kubernetes cluster for development, run:

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/

I suggest reviewing the installation steps, I don't know if this is a bug ... I think that it will be detected in minikube per example

/triage support

**Comment by Prodian0013** [comment]
We have the same issue with n number of nodes on versions 1.14+.  Don't see this issue on 1.13.x.  It's not a pod scheduling issue it seems to be a networking/iptables issues.

**Comment by aojea** [comment]
> We have the same issue with n number of nodes on versions 1.14+. Don't see this issue on 1.13.x. It's not a pod scheduling issue it seems to be a networking/iptables issues.

The iptables rules with the rejects are correct, they are rejecting the traffic because the services don't have endpoints.

The key to understand the problem is, once you boot the cluster and you see those iptables rules, to check the status of the pods and the services, that will give us more info so we can understand the problem. 

I personally couldn't reproduce it but, can you check the status of the flannel and CoreDNS pods when the cluster is rejecting the traffic?

**Comment by Davidrjx** [comment]
@aojea did you resolve coredns problem above? i am also facing NotReadyAddress case for coredns.

> Yeah, seem my theory was wrong 😅
> 
> > NotReadyAddresses: 10.244.0.10,10.244.0.9
> 
> Try to troubleshoot the service and the pods, this is what I see
> 
> ```
> kubectl get pods -n kube-system -l k8s-app=kube-dns
> NAME                       READY   STATUS    RESTARTS   AGE
> coredns-5c98db65d4-6ctb5   1/1     Running   0          27h
> coredns-5c98db65d4-mbxsr   1/1     Running   0          27h
> ```
> 
> ```
> kubectl describe services kube-dns -n kube-systemName:              kube-dns
> Namespace:         kube-system
> Labels:            k8s-app=kube-dns
>                    kubernetes.io/cluster-service=true
>                    kubernetes.io/name=KubeDNS
> Annotations:       prometheus.io/port: 9153
>                    prometheus.io/scrape: true
> Selector:          k8s-app=kube-dns
> Type:              ClusterIP
> IP:                10.96.0.10
> Port:              dns  53/UDP
> TargetPort:        53/UDP
> Endpoints:         10.244.0.2:53,10.244.0.3:53
> Port:              dns-tcp  53/TCP
> TargetPort:        53/TCP
> Endpoints:         10.244.0.2:53,10.244.0.3:53
> Port:              metrics  9153/TCP
> TargetPort:        9153/TCP
> Endpoints:         10.244.0.2:9153,10.244.0.3:9153
> Session Affinity:  None
> Events:            <none>
> ```

**Comment by aojea** [comment]
@Davidrjx I couldn't reproduce the issue, my services are working :sweat_smile:

**Comment by Davidrjx** [comment]
@aojea so your idea or rough steps?

**Comment by Prodian0013** [comment]
> @aojea so your idea or rough steps?

@Davidrjx what's your setup? os, k8s version, cri, networking plugin, etc...

**Comment by omeryounus** [comment]
This issue is still happening. I ran into a problem where I can not access the server. 

While checking the instance console logs, I noticed that the firewall chain KUBE_FIREWALL is blocking all access (Drops packets) to the instance (see below).  

------------------------------------------------------------------------------------------
[ 14.414142] : Chain INPUT (policy DROP)
[ 14.420149] : target prot opt source destination
[ 14.424591] : KUBE-FIREWALL all -- anywhere anywhere

[ 14.815324] : Chain KUBE-FIREWALL (2 references)
[ 14.818313] : target prot opt source destination
[ 14.821858] : DROP all -- anywhere anywhere /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000
[ 14.828311] : DROP all -- !localhost/8 localhost/8 /* block incoming localnet connections */ ! ctstate RELATED,ESTABLISHED,DNAT

When this firewall disable then I connect to the instance and everything else work as normal.
